python/cudf/cudf/utils/cudautils.py

```diff
@@ -16,56 +17,6 @@
 
 
 # Find segments
-
-
-@cuda.jit
-def gpu_mark_found_int(arr, val, out, not_found):
-    i = cuda.grid(1)
-    if i < arr.size:
-        if arr[i] == val:
-            out[i] = i
-        else:
-            out[i] = not_found
-
-
-@cuda.jit
-def gpu_mark_found_float(arr, val, out, not_found):
-    i = cuda.grid(1)
-    if i < arr.size:
-        # TODO: Remove val typecast to float(val)
-        # once numba minimum version is pinned
-        # at 0.51.1, this will have a very slight
-        # performance improvement. Related
-        # discussion in : https://github.com/rapidsai/cudf/pull/6073
-        val = float(val)
-
-        # NaN-aware equality comparison.
-        if (arr[i] == val) or (arr[i] != arr[i] and val != val):
-            out[i] = i
-        else:
-            out[i] = not_found
-
-
-@cuda.jit
-def gpu_mark_gt(arr, val, out, not_found):
-    i = cuda.grid(1)
-    if i < arr.size:
-        if arr[i] > val:
-            out[i] = i
-        else:
-            out[i] = not_found
-
-
-@cuda.jit
-def gpu_mark_lt(arr, val, out, not_found):
-    i = cuda.grid(1)
-    if i < arr.size:
-        if arr[i] < val:
-            out[i] = i
-        else:
-            out[i] = not_found
-
-
 def find_index_of_val(arr, val, mask=None, compare="eq"):
     """
     Returns the indices of the occurrence of *val* in *arr*
```

Comment:
A broader question: why is this function designed like this? The return value isn't what I expect for "Returns the indices of the occurrence of *val* in *arr*".

```python
>>> s = cudf.Series([0, 3, 3, 2, 5, 3])
>>> cudf.utils.cudautils.find_index_of_val(s, 3)
<cudf.core.column.numerical.NumericalColumn object at 0x...>
[
  6,
  1,
  2,
  6,
  6,
  5
]
dtype: int32
```

I expected `[1, 2, 5]` to be the result, not `[6, 1, 2, 6, 6, 5]`. Returning a full column that is the length of the input array indicates that we might have room to improve here.

Reply:
We definitely need to refactor this more broadly, after removing numba. We only use `find_index_of_val` in `find_first` and `find_last`, which are in turn only used in one place each:
- https://github.com/rapidsai/cudf/blob/b8c0501d184a63c59cd70aca189b4887ba4f13fd/python/cudf/cudf/core/column/numerical.py#L611
- https://github.com/rapidsai/cudf/blob/b8c0501d184a63c59cd70aca189b4887ba4f13fd/python/cudf/cudf/core/column/numerical.py#L624

I think we have taken on much more complexity here than we really need.

Reply:
I think these are just fairly old functions for which this is one of the last usages. It's usage in the indexing codepath isn't even right, as it's used to produce results that deviate from pandas in some cases. I expect the function will ultimately be removed entirely as part of @wence- 's broader refactor of indexing, making this PR more of a stopgap as part of the wider MVC effort around organizing numba features. If that's the case then we may as well just wait for the bulk of this to be removed.

Reply:
Great. Removing these entirely is my hope as well.

Reply:
> I expected [1, 2, 5] to be the result, not [6, 1, 2, 6, 6, 5]. Returning a full column that is the length of the input array indicates that we might have room to improve here.

Hmm, yes, agreed. But I am rewriting anyway as Brandon says.